tls/s2n_server_hello.c

```diff
@@ -76,14 +123,17 @@ int s2n_server_hello_recv(struct s2n_connection *conn)
         if (s2n_is_hello_retry_req(conn)) {
             GUARD(s2n_server_hello_retry_recv(conn));
         }
+
         /* Check echoed session ID matches */
         S2N_ERROR_IF(session_id_len != conn->session_id_len || memcmp(session_id, conn->session_id, session_id_len), S2N_ERR_BAD_MESSAGE);
         conn->actual_protocol_version = conn->server_protocol_version;
         GUARD(s2n_set_cipher_as_client(conn, cipher_suite_wire));
     } else {
         uint8_t actual_protocol_version;
-
         conn->server_protocol_version = (uint8_t)(protocol_version[0] * 10) + protocol_version[1];
+
+        S2N_ERROR_IF(s2n_client_detect_downgrade_mechanism(conn), S2N_ERR_SERVER_PROTOCOL_VERSION);
```

Comment:
What do you think about a new s2n_errno for this specific failure? They're cheap :). Something like `S2N_ERR_TLS_DOWNGRADE_DETECTED`? or a better name if you can think of one.

Reply:
Updated in the next push.